### PR TITLE
Ggustavs/feat/python context filter

### PR DESCRIPTION
**Pull request by ggustavs**
Make TensorFlow and PyTorch optional
Make the output of loss function compilation not return the full execution context

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -43,6 +43,31 @@ Each of the following examples comes with an explanatory README file:
 In addition to the above, further examples of specifications can be found in the [test suite](https://github.com/vehicle-lang/vehicle/tree/dev/test/specs)
 and the corresponding output of the Vehicle compiler can be found [here](https://github.com/vehicle-lang/vehicle/tree/dev/test/Test/Compile/Golden).
 
+## Loss backends
+
+The Python package does not install TensorFlow or PyTorch by default so you can
```
Comment (MatthewDaggitt) [inline README.md]: This needs to be added to the `CONTRIBUTING.md` file under `Installing Vehicle` rather than here :+1:  It will then get pulled automatically into the documentation.

File: README.md [hunk 2]
```diff
@@ -43,6 +43,31 @@ Each of the following examples comes with an explanatory README file:
...
+- `pip install "vehicle_lang[pytorch]"`
+
+When using [uv](https://docs.astral.sh/uv/), the equivalent commands are:
+
+```
+uv sync --group tensorflow
+uv sync --group pytorch
+```
+
+Then import the relevant backend module to compile loss functions:
```
Comment (MatthewDaggitt) [inline README.md]: Could you move everything below here to the `docs/training` file?

File: README.md [hunk 3]
```diff
@@ -43,6 +43,31 @@ Each of the following examples comes with an explanatory README file:
...
+
+```python
+from vehicle_lang.loss import tensorflow as loss_tf
+
+losses = loss_tf.load_specification("spec.vcl")
+```
+
+The PyTorch backend works analogously via `vehicle_lang.loss.pytorch`.
```
Comment (MatthewDaggitt) [inline README.md]: Can we make `pytorch` the default for the docs rather than `tensorflow`?

File: vehicle-python/pyproject.toml [hunk 4]
```diff
@@ -101,6 +108,31 @@ testpaths = ["tests"]
...
+mypy = [
+  "types-setuptools>=45",
+  "types-pygments>=2.14,<3",
+]
+pygments = ["pygments>=2.14,<3"]
+numpy = ["numpy>=1.21,<3"]
+jax = ["jax>=0.4.26"]
+dev = ["pre-commit>=4.3.0"]
+tensorflow = [
+  "tensorflow>=2,<3; (python_version >= '3.10' and python_version < '3.13') and ((sys_platform == 'linux' and (platform_machine == 'x86_64' or platform_machine == 'aarch64')) or (sys_platform == 'darwin' and platform_machine == 'x86_64') or (sys_platform == 'win32' and platform_machine == 'AMD64'))",
```
Comment (MatthewDaggitt) [inline vehicle-python/pyproject.toml]: Do we really need to duplicate all this stuff here? If so, can you add a comment explaining why?

File: vehicle-python/src/vehicle_lang/loss/_abc/_translation.py [hunk 5]
```diff
@@ -1,49 +1,33 @@
...
-from vehicle_lang.compile._abc import _types as vcl_var
-
 # from ... import ast as vcl_ast
-from vehicle_lang.compile._ast import _nodes as vcl_ast
+from .._ast import _nodes as vcl_ast
+from . import _types as vcl_var
 
 ################################################################################
-### Translation from Vehicle AST to Python AST
+### Translation from vehicle_lang AST to Python AST
```
Comment (MatthewDaggitt) [inline vehicle-python/src/vehicle_lang/loss/_abc/_translation.py]: I think this should stay `Vehicle`?

File: vehicle-python/src/vehicle_lang/loss/_abc/_translation.py [hunk 6]
```diff
@@ -1,49 +1,33 @@
 from abc import ABCMeta, abstractmethod
 from typing import Generic
 
-from typing_extensions import override
-from vehicle_lang.compile._abc import _types as vcl_var
-
 # from ... import ast as vcl_ast
```
Comment (MatthewDaggitt) [inline vehicle-python/src/vehicle_lang/loss/_abc/_translation.py]: Can delete?

File: vehicle-python/src/vehicle_lang/loss/_python/_translation.py [hunk 7]
```diff
@@ -59,12 +73,12 @@ def compile(
     def translate_Main(self, program: vcl.Main) -> py.Module:
         return py.Module(
             body=[
-                # NOTE: 'vehicle_lang.ast' is imported for 'Tensor'
+                # NOTE: 'vehicle.ast' is imported for 'Tensor'
```
Comment (MatthewDaggitt) [inline vehicle-python/src/vehicle_lang/loss/_python/_translation.py]: Should be `vehicle_lang`?

**Review by MatthewDaggitt — CHANGES_REQUESTED** [review]

**Comment by MatthewDaggitt** [comment]
Also in this PR, could you add a note to the CHANGELOG that we've dropped support for certain Python versions?

File: README.md (hunk 3 above)
Comment (ggustavs) [reply README.md]: Done. Will commit when you answer in Slack :)

File: vehicle-python/src/vehicle_lang/loss/_abc/_translation.py (hunk 5 above)
Comment (ggustavs) [reply vehicle-python/src/vehicle_lang/loss/_abc/_translation.py]: Yup, consequences of not making my search and replace case-sensitive :|

File: vehicle-python/src/vehicle_lang/loss/_abc/_translation.py (hunk 6 above)
Comment (ggustavs) [reply vehicle-python/src/vehicle_lang/loss/_abc/_translation.py]: Yup, look below, different import

File: vehicle-python/pyproject.toml (hunk 4 above)
Comment (ggustavs) [reply vehicle-python/pyproject.toml]: Removed, you are right :)

File: docs/dev/unstable-python-api.rst [hunk 8]
```diff
@@ -190,3 +37,12 @@ Unstable Python API
...
    :undoc-members:
+
+Legacy modules
+--------------
+
+Earlier releases exposed ``vehicle_lang.ast`` and ``vehicle_lang.compile.*`` helper
+modules. The modern Python bindings now marshal data directly through the Vehicle
+compiler, so those legacy interfaces are no longer published. Consult the git
+history (v0.20 and earlier) if you need the old AST helpers for research or
+migration work.
```
Comment (MatthewDaggitt) [inline docs/dev/unstable-python-api.rst]: This whole file can just be nuked. No one is using the old Python API. I would like to move the Python bindings documentation under the documentation for each correct specific mode.

File: docs/python-api.rst [hunk 9]
```diff
@@ -1,34 +1,91 @@
 Python API
 ==========
```
Comment (MatthewDaggitt) [inline docs/python-api.rst]: See comment above. Can we get rid of this file, and merge this documentation into the submodules in the appropriate place, e.g. `Installation` into `CONTRIBUTING.md`, the stuff about loss functions into `training.rst` etc?

**Comment by ggustavs** [comment]
The tests are failing cause dockerhub is down (the side serving push and pull requests)

File: docs/training.rst [hunk 10]
```diff
@@ -1,7 +1,125 @@
...
 =============================
+Vehicle compiles each ``@property`` in your specification into a callable loss function. Use one of the backend modules (PyTorch or TensorFlow) to load the compiled declarations into Python.
 
-.. warning::
+Backend extras
+--------------
 
-    The Python bindings are not yet documented, apologies.
-    See the [Vehicle tutorial](https://vehicle-lang.github.io/tutorial/) for details of how to use them.
+The `vehicle_lang` wheel keeps deep-learning frameworks behind optional extras so you only install what you need:
```
Comment (MatthewDaggitt) [inline docs/training.rst]: Can we change `wheel` to `package`?

File: docs/training.rst [hunk 11]
```diff
@@ -1,7 +1,125 @@
...
+
+.. code-block:: bash
+
+    pip install "vehicle_lang[pytorch]"
+    pip install "vehicle_lang[tensorflow]"
+
+Install both extras (or combine them with ``test``/``pygments``) if you need to exercise several backends in the same environment.
```
Comment (MatthewDaggitt) [inline docs/training.rst]: Can we move this to the `installation.md` section rather than here? We can back-reference if we want but users should first see this information at install time rather than here...

File: docs/training.rst [hunk 12]
```diff
@@ -1,7 +1,125 @@
...
+
+Loading declarations
```
Comment (MatthewDaggitt) [inline docs/training.rst:6]: `Generating a loss function`?

File: docs/training.rst [hunk 13]
```diff
@@ -1,7 +1,125 @@
...
+   context = {"model": model, "epsilon": 0.1}
+
+   declarations = loss_pt.load_specification(
+       "spec.vcl",
+       samplers=custom,
+       declaration_context=context,
+       declarations=["output_bounded"],
+   )
+
+TensorFlow works the same way—load the declarations through ``vehicle_lang.loss.tensorflow`` and call the compiled property inside a ``tf.GradientTape`` so gradients flow through both the task loss and the constraint loss:
```
Comment (MatthewDaggitt) [inline docs/training.rst:56]: This feels like it should go above the previous section talking about samplers?

File: docs/training.rst [hunk 14]
```diff
@@ -1,7 +1,125 @@
...
+
+   grads = tape.gradient(total_loss, model.trainable_variables)
+   optimizer.apply_gradients(zip(grads, model.trainable_variables))
+
+By combining your usual task loss with the constraint losses returned by Vehicle, you ensure the optimiser simultaneously keeps the model on task and enforces the specification.
+
+Logic selection
+---------------
+
+Switch the differentiable backend by passing ``logic=`` when calling ``load_specification``. ``DifferentiableLogic.DL2`` matches the classic loss semantics, while ``DifferentiableLogic.Vehicle`` enables the experimental optimisation pipeline that keeps compilation inside Vehicle whenever possible.
```
Comment (MatthewDaggitt) [inline docs/training.rst]: This sentence is pretty much nonsense. I would just provide a very short introduction to the concept of differentiable logics and then list the currently available options.

File: docs/training.rst [hunk 15]
```diff
@@ -1,7 +1,125 @@
...
+
+           total_loss.backward()
+           optimizer.step()
+
+Here ``train_loader`` can be any ``torch.utils.data.DataLoader`` that yields ``(x_batch, y_batch)`` pairs, and the ``network`` callable must match the type you declared for the Vehicle network so gradient-based samplers can evaluate it.
+
+Custom samplers and declaration context
+---------------------------------------
+
+Pass a ``samplers`` dictionary if you want to override how individual properties explore the search space. Each entry maps the ``@property`` name to a callable with the ``get_loss`` signature described in ``vehicle_lang.loss._common``. ``declaration_context`` lets you share state across declarations, for example to reuse an optimiser or cache dataset handles.
```
Comment (MatthewDaggitt) [inline docs/training.rst]: I think this section needs to come below the `Differentiable logics` section, and have a couple of introductory sentences on how quantifiers are translated.

**Review by MatthewDaggitt — CHANGES_REQUESTED** [review]
Other than the comments on the documentation, is this ready to merge @ggustavs?

File: docs/training.rst (hunk 12 above)
Comment (wenkokke) [reply docs/training.rst:6]: Part of the point is that pygments is also an optional extra.

File: docs/training.rst (hunk 10 above)
Comment (ggustavs) [reply docs/training.rst]: Done

**Comment by ggustavs** [comment]
> Other than the comments on the documentation, is this ready to merge @ggustavs?

Yes, everything else is ready. 

Mb about the docs - was quite sleep deprived when I wrote them :|

**Review by MatthewDaggitt — APPROVED** [review]
